test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,410 @@
+  }
+  return tmp_s;
+}
+
+std::shared_ptr<EventEngine> GetPosixEE() {
+  static std::shared_ptr<EventEngine> posix_ee =
+      std::make_shared<PosixEventEngine>();
+  return posix_ee;
+}
+
+EventEngine* GetOracleEE() {
+  static EventEngine* oracle_ee = new PosixOracleEventEngine();
+  return oracle_ee;
+}
+
+class TestScheduler : public Scheduler {
+ public:
+  explicit TestScheduler(EventEngine* engine) : engine_(engine) {}
+  void Run(EventEngine::Closure* closure) override { engine_->Run(closure); }
+
+  void Run(absl::AnyInvocable<void()> cb) override {
+    engine_->Run(std::move(cb));
+  }
+
+ private:
+  EventEngine* engine_;
+};
+
+std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                     std::unique_ptr<EventEngine::Endpoint>>>
+CreateConnectedEndpoints(PosixEventPoller* poller, bool is_zero_copy_enabled,
+                         int num_connections) {
+  EXPECT_NE(GetOracleEE(), nullptr);
+  EXPECT_NE(GetPosixEE(), nullptr);
+  EXPECT_NE(poller, nullptr);
+  std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                       std::unique_ptr<EventEngine::Endpoint>>>
+      connections;
+  auto memory_quota = absl::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  EventEngine::ResolvedAddress resolved_addr =
+      URIToResolvedAddress(target_addr);
+  Promise<std::unique_ptr<EventEngine::Endpoint>> server_endpoint_promise;
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint_promise](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint_promise.Set(std::move(ep));
+      };
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  if (is_zero_copy_enabled) {
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_ENABLED, 1);
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_SEND_BYTES_THRESHOLD,
+                    kMinMessageSize);
+  }
+  ChannelArgsEndpointConfig config(args);
+  auto status = GetOracleEE()->CreateListener(
+      std::move(accept_cb),
+      [](absl::Status status) { ASSERT_TRUE(status.ok()); }, config,
+      absl::make_unique<grpc_core::MemoryQuota>("foo"));
+  EXPECT_TRUE(status.ok());
+
+  std::unique_ptr<Listener> listener = std::move(*status);
+  EXPECT_TRUE(listener->Bind(resolved_addr).ok());
+  EXPECT_TRUE(listener->Start().ok());
+
+  // Create client socket and connect to the target address.
+  for (int i = 0; i < num_connections; ++i) {
+    int client_fd;
+    int one = 1;
+    int flags;
+
+    client_fd = socket(AF_INET6, SOCK_STREAM, 0);
+    setsockopt(client_fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
+    // Make fd non-blocking.
+    flags = fcntl(client_fd, F_GETFL, 0);
+    EXPECT_EQ(fcntl(client_fd, F_SETFL, flags | O_NONBLOCK), 0);
+
+    if (connect(client_fd,
```

Comment:
This appears to be the third copy of this code in the codebase. Please deduplicate.

Reply:
Moved it to posix_engine_test_utils{.h,.cc} files into a function called ConnectToServerOrDie